The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11336748375/job/31527188296) [(plain)](https://github.com/rust-lang/rust/commit/e0fb0dcfbc8f4078fb7d829b1cfb071b30ac6a9f/checks/31527188296/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.916 Building wheels for collected packages: reuse
#13 2.917   Building wheel for reuse (pyproject.toml): started
#13 3.164   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.165   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.165   Stored in directory: /tmp/pip-ephem-wheel-cache-d90k1du8/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.168 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.556 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.557 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.090 Collecting virtualenv
#13 4.090 Collecting virtualenv
#13 4.143   Downloading virtualenv-20.26.6-py3-none-any.whl (6.0 MB)
#13 4.375      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 26.1 MB/s eta 0:00:00
#13 4.434 Collecting platformdirs<5,>=3.9.1
#13 4.442   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.464 Collecting distlib<1,>=0.3.7
#13 4.483      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 53.0 MB/s eta 0:00:00
#13 4.521 Collecting filelock<4,>=3.12.2
#13 4.529   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.529   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.610 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.799 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.6
#13 DONE 4.9s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      237504 kB
DirectMap2M:     9199616 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARNING: `rust.download-rustc` is enabled. The `rust.channel` option will be overridden by the CI rustc's channel.
downloading https://static.rust-lang.org/dist/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
                         report_diagnostic(
                             self.cx.tcx,
                             BROKEN_INTRA_DOC_LINKS,
-                            format!(
-                                "all items matching `{path_str}` are private or doc(hidden)"
-                            ),
+                            format!("all items matching `{path_str}` are private or doc(hidden)"),
                             &diag_info,
                             |diag, sp, _| {
                                 if let Some(sp) = sp {
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustdoc/passes/strip_priv_imports.rs" "/checkout/tests/run-make/print-cfg/rmake.rs" "/checkout/tests/run-make/rustdoc-with-output-option/rmake.rs" "/checkout/tests/run-make/rustdoc-with-output-option/src/lib.rs" "/checkout/src/librustdoc/passes/lint/unescaped_backticks.rs" "/checkout/src/librustdoc/passes/lint/html_tags.rs" "/checkout/src/librustdoc/passes/lint/check_code_block_syntax.rs" "/checkout/src/librustdoc/passes/lint/redundant_explicit_links.rs" "/checkout/src/librustdoc/passes/lint/bare_urls.rs" "/checkout/src/librustdoc/passes/lint/unportable_markdown.rs" "/checkout/src/librustdoc/passes/propagate_stability.rs" "/checkout/tests/run-make/wasm-custom-section/rmake.rs" "/checkout/tests/run-make/wasm-custom-section/foo.rs" "/checkout/tests/run-make/wasm-custom-section/bar.rs" "/checkout/src/librustdoc/passes/propagate_doc_cfg.rs" "/checkout/src/librustdoc/passes/collect_trait_impls.rs" "/checkout/src/librustdoc/passes/strip_private.rs" "/checkout/src/librustdoc/passes/lint.rs" "/checkout/src/librustdoc/passes/check_doc_test_visibility.rs" "/checkout/src/librustdoc/passes/strip_aliased_non_local.rs" "/checkout/src/librustdoc/passes/collect_intra_doc_links.rs" "/checkout/src/librustdoc/passes/strip_hidden.rs" "/checkout/src/librustdoc/passes/calculate_doc_coverage.rs" "/checkout/src/librustdoc/passes/stripper.rs" "/checkout/src/librustdoc/passes/mod.rs" "/checkout/tests/run-make/rustdoc-cfgspec-parsing/rmake.rs" "/checkout/tests/run-make/rustdoc-cfgspec-parsing/foo.rs" "/checkout/tests/run-make/raw-dylib-alt-calling-convention/driver.rs" "/checkout/tests/run-make/raw-dylib-alt-calling-convention/rmake.rs" "/checkout/tests/run-make/raw-dylib-alt-calling-convention/lib.rs" "/checkout/src/librustdoc/clean/cfg.rs" "/checkout/src/librustdoc/clean/types.rs" "/checkout/src/librustdoc/clean/inline.rs" "/checkout/tests/run-make/c-unwind-abi-catch-panic/main.rs" "/checkout/tests/run-make/c-unwind-abi-catch-panic/rmake.rs" "/checkout/src/librustdoc/clean/utils/tests.rs" "/checkout/src/librustdoc/clean/blanket_impl.rs" "/checkout/src/librustdoc/clean/simplify.rs" "/checkout/tests/run-make/cross-lang-lto-riscv-abi/rmake.rs" "/checkout/tests/run-make/cross-lang-lto-riscv-abi/riscv-xlto.rs" "/checkout/src/librustdoc/clean/types/tests.rs" "/checkout/src/librustdoc/clean/render_macro_matchers.rs" "/checkout/src/librustdoc/clean/auto_trait.rs" "/checkout/src/librustdoc/clean/utils.rs" "/checkout/tests/run-make/naked-symbol-visibility/rmake.rs" "/checkout/tests/run-make/naked-symbol-visibility/a_rust_dylib.rs" "/checkout/src/librustdoc/clean/cfg/tests.rs" "/checkout/src/librustdoc/clean/mod.rs" "/checkout/src/librustdoc/markdown.rs" "/checkout/tests/run-make/extern-fn-slice-no-ice/rmake.rs" "/checkout/tests/run-make/extern-fn-slice-no-ice/test.rs" "/checkout/src/librustdoc/json/import_finder.rs" "/checkout/src/librustdoc/json/conversions.rs" "/checkout/src/librustdoc/json/mod.rs" "/checkout/src/librustdoc/doctest.rs" "/checkout/src/librustdoc/scrape_examples.rs" "/checkout/src/librustdoc/docfs.rs" "/checkout/src/librustdoc/fold.rs" "/checkout/src/librustdoc/config.rs" "/checkout/src/librustdoc/core.rs" "/checkout/src/librustdoc/externalfiles.rs" "/checkout/src/librustdoc/theme.rs" "/checkout/src/librustdoc/visit.rs" "/checkout/tests/run-make/lto-no-link-whole-rlib/lib2.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Tue Oct 15 00:05:56 UTC 2024
  network time: Tue, 15 Oct 2024 00:05:56 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/src/librustdoc/passes/collect_intra_doc_links.rs
+++ b/src/librustdoc/passes/collect_intra_doc_links.rs
@@ -288,14 +288,16 @@ pub(crate) struct LinkCollector<'a, 'tcx> {
     /// Cache the resolved links so we can avoid resolving (and emitting errors for) the same link.
     /// The link will be `None` if it could not be resolved (i.e. the error was cached).
     pub(crate) visited_links: FxHashMap<ResolutionInfo, Option<(Res, Option<UrlFragment>)>>,
-    /// These links are ambiguous. We need for the cache to have its paths filled. Unfortunately,
-    /// if we run the `LinkCollector` pass after `Cache::populate`, a lot of items that we need
-    /// to go through will be removed, making a lot of intra-doc links to not be inferred.
+    /// According to `rustc_resolve`, these links are ambiguous.
     ///
-    /// So instead, we store the ambiguous links and we wait for cache paths to be filled before
-    /// inferring them (if possible).
+    /// However, we cannot link to an item that has been stripped from the documentation. If all
+    /// but one of the "possibilities" are stripped, then there is no real ambiguity. To determine
+    /// if an ambiguity is real, we delay resolving them until after `Cache::populate`, then filter
+    /// every item that doesn't have a cached path.
     ///
-    /// Key is `(item ID, path str)`.
+    /// We could get correct results by simply delaying everything. This would have fewer happy
+    /// codepaths, but we want to distinguish different kinds of error conditions, and this is easy
+    /// to do by resolving links as soon as possible.
     pub(crate) ambiguous_links: FxIndexMap<(ItemId, String), Vec<AmbiguousLinks>>,
 }
 
@@ -1187,9 +1189,7 @@ impl LinkCollector<'_, '_> {
                         report_diagnostic(
                             self.cx.tcx,
                             BROKEN_INTRA_DOC_LINKS,
-                            format!(
-                                "all items matching `{path_str}` are either private or doc(hidden)"
-                            ),
+                            format!("all items matching `{path_str}` are private or doc(hidden)"),
                             &diag_info,
                             |diag, sp, _| {
                                 if let Some(sp) = sp {

--- a/src/librustdoc/passes/collect_intra_doc_links.rs
+++ b/src/librustdoc/passes/collect_intra_doc_links.rs
@@ -302,7 +302,6 @@ pub(crate) struct LinkCollector<'a, 'tcx> {
 }
 
 pub(crate) struct AmbiguousLinks {
-    disambiguator: Option<Disambiguator>,
     link_text: Box<str>,
     diag_info: OwnedDiagnosticInfo,
     resolved: Vec<(Res, Option<UrlFragment>)>,
@@ -1097,7 +1096,6 @@ impl LinkCollector<'_, '_> {
 
         if resolved.len() > 1 {
             let links = AmbiguousLinks {
-                disambiguator,
                 link_text: link_text.clone(),
                 diag_info: diag_info.into(),
                 resolved,
@@ -1178,7 +1176,7 @@ impl LinkCollector<'_, '_> {
                             res,
                             fragment,
                             path_str,
-                            info.disambiguator,
+                            None,
                             diag_info,
                             &info.link_text,
                         ) {

--- a/tests/rustdoc-ui/intra-doc/filter-out-private-2.rs
+++ b/tests/rustdoc-ui/intra-doc/filter-out-private-2.rs
@@ -11,5 +11,5 @@ pub struct Thing {}
 pub fn Thing() {}
 
 /// Do stuff with [`Thing`].
-//~^ ERROR all items matching `Thing` are either private or doc(hidden)
+//~^ ERROR all items matching `Thing` are private or doc(hidden)
 pub fn repro(_: Thing) {}

--- a/tests/rustdoc-ui/intra-doc/filter-out-private-2.stderr
+++ b/tests/rustdoc-ui/intra-doc/filter-out-private-2.stderr
@@ -1,4 +1,4 @@
-error: all items matching `Thing` are either private or doc(hidden)
+error: all items matching `Thing` are private or doc(hidden)
   --> $DIR/filter-out-private-2.rs:13:21
    |
 LL | /// Do stuff with [`Thing`].

